File: Kotlin/src/main/kotlin/chapter5_command/RemoteControl.kt
package chapter5_command

import chapter5_command.command.Command
import chapter5_command.command.NoCommand
import java.util.Stack

class RemoteControl {
    private val onCommands = mutableListOf<Command>()
    private val offCommands = mutableListOf<Command>()
    private var undoCommands = Stack<Command>()

    init {
        val noCommand = NoCommand()
        for (i: Int in 1 .. SLOT_SIZE) {
            onCommands.add(noCommand)
            offCommands.add(noCommand)
        }
    }

    fun setCommand(slot: Int, onCommand: Command, offCommand: Command) {
        onCommands[slot] = onCommand
        offCommands[slot] = offCommand
    }

    fun onButtonWasPushed(slot: Int) {
        onCommands[slot].execute()
        undoCommands.push(onCommands[slot])
    }

    fun offButtonWasPushed(slot: Int) {
        offCommands[slot].execute()
        undoCommands.push(offCommands[slot])
    }

    fun undoButtonWasPushed() {
        if (undoCommands.size > 0) {
            val cmd = undoCommands.pop()
            cmd.undo()
        } else {
            println("취소할 명령이 없습니다")
        }
    }

    companion object {
        const val SLOT_SIZE = 7
    }
}

File: Scala/src/main/scala/command/RemoteControl.scala
package command

import scala.collection.mutable

class RemoteControl {

  val controlSize = 7
  val commands = new Array[Command](controlSize)
  val noCommand = new NoCommand()
  for (i <- 0 until 7) {
    commands(i) = noCommand
  }
  val undoCommands = new mutable.ListBuffer[() => Unit]

  def setCommand(slot: Int, command: Command): Unit = {
    commands(slot) = command
  }

  def onButtonWasPushed(slot: Int): Unit = {
    commands(slot).execute()
    undoCommands.append(commands(slot).stop)
  }

  def offButtonWasPushed(slot: Int): Unit = {
    commands(slot).stop()
    undoCommands.append(commands(slot).execute)
  }

  def undo(): Unit = {
    if (undoCommands.isEmpty) {
      println("No Command to execute undo")
    } else {
      val undoCommand = undoCommands.last
      undoCommands.remove(undoCommands.length - 1)
      undoCommand()
    }
  }

  override def toString: String = {
    commands.zipWithIndex.map(x => {
      s"${x._2} - ${x._1.toString}"
    }).mkString("\n")
  }
}
